### PR TITLE
new firmware version 9

### PR DESCRIPTION
**Pull request by mbonani**
correct some bug in native function
improve disconnection
add setting mode

### CODE REVIEW

File: behavior.c [hunk 1]
```diff
@@ -371,33 +372,246 @@ static void behavior_leds_ntc(void ) {
...
+	unsigned char ledrgb[3];
 
 	all_off = !(buttons_state[0] | buttons_state[1] | buttons_state[2]
-		| buttons_state[3] | buttons_state[4]);
-	
-	if(buttons_state[1] && buttons_state[4]) {
-		if(++dbnc > 100) {
-			if(dbnc == 101) {
-				if(!(rf_get_status() & RF_PAIRING_MODE)) {
+			| buttons_state[3] | buttons_state[4]);
```
Comment (retornaz) [inline behavior.c]: ???

File: main.c [hunk 2]
```diff
@@ -164,6 +165,7 @@ static void timer_slow(void) {
 }
 
 static unsigned int timer[2];
+extern unsigned int reconnection_delay;
```
Comment (retornaz) [inline main.c]: ????

File: mode.c [hunk 3]
```diff
@@ -57,13 +57,7 @@ static unsigned int bs_black_level = 260;
...
 
 
-static void set_body_rgb(unsigned int r, unsigned int g, unsigned int b) {
-	leds_set_top(r,g,b);
-	leds_set_br(r,g,b);
-	leds_set_bl(r,g,b);
-}
-
-static int pulse_get(void) {
+int pulse_get(void) {
```
Comment (retornaz) [inline mode.c]: static ?

File: behavior.c [hunk 4]
```diff
@@ -371,33 +372,246 @@ static void behavior_leds_ntc(void ) {
...
+		if (usb_uart_serial_port_open() || (rf_get_status() & RF_DATA_RX)) {
+			behavior_stop(B_SETTING);
+			leds_set_body_rgb(0, 0, 0);
+			leds_set_circle(0, 0, 0, 0, 0, 0, 0, 0);
+			leds_set_prox_h(0, 0, 0, 0, 0, 0, 0, 0);
+			vmVariables.target[0] = 0;
+			vmVariables.target[1] = 0;
+			if (rf_get_status() & RF_PAIRING_MODE)
+				rf_pairing_stop();
+			init_vm_mode();
```
Comment (retornaz) [inline behavior.c]: It's a layering violation. But if it works  ...

File: skel-usb.c [hunk 5]
```diff
@@ -38,6 +38,7 @@
 
 #include "skel-usb.h"
 
+unsigned char update_calib;
```
Comment (retornaz) [inline skel-usb.c]: ... I told you already to NOT add any global variables ... 
Add a function which will _SET_ to one the variable when needed.
Yes it looks a bit "overkill" but this force the code to only set this variable.

File: thymio-buffer.c [hunk 6]
```diff
@@ -121,8 +121,8 @@ unsigned char AsebaFifoTxPeek(void) {
 
 int AsebaFifoTxEmpty(void) {
 	return !get_used(&AsebaFifo.tx);
-}	 
-
+}	
+extern struct _vmVariables vmVariables;
```
Comment (retornaz) [inline thymio-buffer.c]: WTF ?!

**Comment by retornaz** [comment]
We are close to have something mergeable
Please use git rebase -i to remove the unneeded commit (reconnection delay add + revert, several .hex add+remove) and squash together the code add + cleaning. This will make history much much more cleaner: https://robots.thoughtbot.com/git-interactive-rebase-squash-amend-rewriting-history

File: behavior.c (hunk 1 above)
Comment (mbonani) [inline behavior.c]: the diff becames strange when I supress the pairing tick because it used same piece of code...

Comment (retornaz) [inline behavior.c]: the comment is not about the look of the diff, but about the all_off variable.
Please explain why the variable is still present.

**Comment by mbonani** [comment]
Hi,
I correct the code following your suggestion and clean the history with rebase. 
Cheers

**Comment by retornaz** [comment]
It's getting better ! 

Please squash:

- "fix disconnection bug" into "DTR/DTR connection problem"
- "restruck to remove update_calib as global variabl" and "re-architecture code to have a central point managing settings" into "flash setting when modify"
- "rename function pulse_get()" into "add volume control setting mode"
- "remove pairing behavior, its is include in setting behavior" into "add setting mode structure"
- "remove unused variable " into "add setting mode structure "

Remove the "format" and the revert of it (squash if not possible, with "small esthetic" and "minor esthetic changes")).
If the remove of format and the revert is possible, squash the "small esthetic" and "minor esthetic changes" together.

After this, you will be a git rebase master :)

**Comment by mbonani** [comment]
I mostly manage to make the rebase asked. (it was painful)
There is just "rename function pulse_get()" that will bea mess to squash it into volume control... I propose to squash it into "change the visual of setting mode.
I accept status of "git rebase master", could you provide me a certification? ;-)

**Comment by retornaz** [comment]
You'll be a git rebase PhD when you'll manage to rebase + fixup the commits after in history, but for now Master is enough to merge this pull request :p

I know it was painful for you, but I think you master a bit more git now which can only be positive.

**Comment by mbonani** [comment]
Thank you to show me the way to the force of git, master

Padavan Michael

Excuse my brevity sent from my smartphone
Le 29 mai 2015 19:48, "Philippe Rétornaz" <notifications@github.com> a
écrit :

**Pull request merged**